### PR TITLE
ARROW-7306: [C++] Add Result-returning version of FileSystemFromUri

### PR DESCRIPTION
**Pull request by mrkn**
I guess this should be in #5866

### CODE REVIEW

File: cpp/src/arrow/filesystem/filesystem.h [hunk 1]
```diff
@@ -325,9 +337,13 @@ class ARROW_EXPORT SlowFileSystem : public FileSystem {
...
 /// \param[out] out_path (optional) Path inside the filesystem.
 /// \return Status
-ARROW_EXPORT
-Status FileSystemFromUri(const std::string& uri, std::shared_ptr<FileSystem>* out_fs,
-                         std::string* out_path = NULLPTR);
+ARROW_DEPRECATED("Use Result-returning version")
+inline Status FileSystemFromUri(const std::string& uri,
+                                std::shared_ptr<FileSystem>* out_fs,
+                                std::string* out_path = NULLPTR) {
+  ARROW_ASSIGN_OR_RAISE(*out_fs, FileSystemFromUri(uri, out_path));
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/filesystem.h]: You can simply write `return FileSystemFromUri(uri, out_path).Value(out_fs);`

**Review by pitrou — COMMENTED** [review]
LGTM, just one nit.

File: cpp/src/arrow/filesystem/filesystem.h (hunk 1 above)
Comment (mrkn) [reply cpp/src/arrow/filesystem/filesystem.h]: Oops, I missed that feature!

**Review by pitrou — APPROVED** [review]
+1. Thank you @mrkn !

**Pull request closed without merging**